src/Chainweb/Pact/PactService/ExecBlock.hs

```diff
@@ -148,12 +148,9 @@ execBlock currHeader plData pdbenv = do
 
     modify' $ set psStateValidated $ Just currHeader
 
-    -- Validate hashes if requested
-    asks _psValidateHashesOnReplay >>= \x -> when x $
-        either throwM (void . return) $!
-        validateHashes currHeader plData miner results
-
-    return $! T2 miner results
+    -- validate hashes only on evaluation of the result
+    return $
+      validateHashes currHeader plData miner results
```

Comment:
As described above, I think, it is cleaner to keep using `_psValidateHashesOnReplay` (with `True` as default value, but could also be the other way around) and set it locally in the environment as needed.

The return type could then be simplified to just `PactServiceM tbl PayloadWithOutputs`, because all calls to `validateHashes` and possibly throwing an exception would happen here in execBlock.

Reply:
The invariant here is that we should not produce a `PayloadWithOutputs` without validating it, because the reason we would need a `PayloadWithOutputs` is to insert it into the database. So we have two possible contexts here: either
1. we're validating a block we've already validated, so we don't need an artifact we can insert into the database and we're doing it purely for the effect on the checkpointer and Pact state, so the result can be ignored, or
2. we're validating a new block, so we do need an artifact we can insert into the database.

To preserve this invariant, if we remove the laziness here, we can either
1. validate unconditionally, which requires that we analyze the performance of validation, or
2. stop returning a `PayloadWithOutputs` and instead return what we used to return, the pair of `(Miner, Transactions)`, which we can make a `PayloadWithOutputs` from. My worry there is that the caller can choose to just make a `PayloadWithOutputs` without validating it, and it is as far as I know never valid to make a `PayloadWithOutputs` from the output of this function without validating.

So the code shown is Pareto-optimal in a sense, it both prevents us from shooting ourselves in the foot by producing an invalid `PayloadWithOutputs`, and it prevents us from doing the work of validating that `PayloadWithOutputs` if we don't need to produce it at all.

One other thing we can do, is we can make this function return `IO (IO PayloadWithOutputs)`, and make executing the inner IO action trigger hash validation. To me, though, this seems even more confusing, because it's not obvious which effects exactly are in the inner action, whereas it's intuitive to me that an `Either PactException PayloadWithOutputs` will when evaluated do exactly the work necessary to produce a valid `PayloadWithOutputs`.

Reply:
> The invariant here is that we should not produce a PayloadWithOutputs without validating it

Why not make that explicit on the type level by returning `Maybe PayloadWithOutputs` or using two functions (different semantics deserves different functions)? Using laziness is completely invisible on the type level but still affects semantics (in this time mostly perf).